repo/content/index_blob_manager.go

```diff
@@ -221,63 +221,86 @@ func (m *indexBlobManagerImpl) deleteOldIndexBlobs(ctx context.Context, latestBl
 
 	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
 	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
-	latestServerBlobTime := latestBlob.Timestamp
+	compactionLogServerTimeCutoff := latestBlob.Timestamp.Add(-m.minIndexBlobDeleteAge)
+	compactionBlobs := blobsOlderThan(allCompactionLogBlobs, compactionLogServerTimeCutoff)
 
-	compactionLogServerTimeCutoff := latestServerBlobTime.Add(-m.minIndexBlobDeleteAge)
-	fetchCompactionLogs := blobsOlderThan(allCompactionLogBlobs, compactionLogServerTimeCutoff)
-	log(ctx).Debugf("fetching %v/%v compaction logs older than %v", len(fetchCompactionLogs), len(allCompactionLogBlobs), compactionLogServerTimeCutoff)
+	log(ctx).Debugf("fetching %v/%v compaction logs older than %v", len(compactionBlobs), len(allCompactionLogBlobs), compactionLogServerTimeCutoff)
 
-	entries, err := m.getCompactionLogEntries(ctx, fetchCompactionLogs)
+	compactionBlobEntries, err := m.getCompactionLogEntries(ctx, compactionBlobs)
 	if err != nil {
 		return errors.Wrap(err, "unable to get compaction log entries")
 	}
 
-	indexBlobsToDelete := map[blob.ID]bool{}
+	indexBlobsToDelete := m.findIndexBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobEntries)
+	compactionLogBlobsToDelete := m.findCompactionLogBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobs)
+
+	// note that we must always delete index blobs first before compaction logs
+	// otherwise we may inadvertedly resurrect an index blob that should have been removed.
+	if err := m.deleteBlobsFromStorageAndCache(ctx, indexBlobsToDelete); err != nil {
+		return errors.Wrap(err, "unable to delete compaction logs")
+	}
+
+	if err := m.deleteBlobsFromStorageAndCache(ctx, compactionLogBlobsToDelete); err != nil {
+		return errors.Wrap(err, "unable to delete compaction logs")
+	}
+
+	m.flushCache()
+
+	return nil
+}
+
+func (m *indexBlobManagerImpl) findIndexBlobsToDelete(ctx context.Context, latestServerBlobTime time.Time, entries map[blob.ID]*compactionLogEntry) []blob.ID {
+	tmp := map[blob.ID]bool{}
 
 	for _, cl := range entries {
 		for _, b := range cl.InputBlobs {
 			if age := latestServerBlobTime.Sub(b.Timestamp); age < m.minIndexBlobDeleteAge {
-				log(ctx).Debugf("not deleting index blob %v, because it's too recent: %v < %v", b.BlobID, age, m.minIndexBlobDeleteAge)
+				log(ctx).Debugf("not deleting compacted index blob %v, because it's too recent: %v < %v", b.BlobID, age, m.minIndexBlobDeleteAge)
```

Comment:
Somewhere, before an index compaction process starts, the `minIndexBlobDeleteAge` and `minCompactionLogBlobDeleteAge` values must be validated to ensure that `minIndexBlobDeleteAge < minCompactionLogBlobDeleteAge`. This is to prevent a situation where a (relatively recent) compaction log is chosen for deletion, but the corresponding index blobs are not deleted because they happen to be too recent w.r.t. `minIndexBlobDeleteAge`.
If that has been validated upfront, then this would become an "impossibility" here, and the safe thing to do would be to return an error in that case.

Reply:
it's already there - first line of `deleteOldIndexBlobs()`